source/app/controllers/urls_controller.rb

```diff
@@ -1,2 +1,47 @@
 class UrlsController < ApplicationController
+  # GET /urls
+  # GET /urls.json
+  def index
+    redirect_to login_path and return if !logged_in?
+    @urls = Url.all
+  end
+
+  # GET /urls/1
+  # GET /urls/1.json
+  def show
+    url = Url.find_by(short_url: params[:id])
```

Comment:
Sorry I wasn't clear. Rails rescues `ActiveRecord::RecordNotFound` and responds with 404 in production, not development or test.